plutus-conformance/add-test-output/Spec.hs

```diff
@@ -61,58 +61,76 @@ runnerReader inp =
-            case parsed of
-              Left (ParseErrorB peb) -> do
-                -- warn the user that the file failed to parse
-                putStrLn $ inputFile <> " failed to parse. Error written to " <> outFilePath
-                writeFile outFilePath (errorBundlePretty peb)
-              Right pro -> do
-                res <- try (evaluate $ evalUplcProg (() <$ pro)):: IO (Either SomeException (EvaluationResult UplcProg))
-                case res of
-                  Right (EvaluationSuccess prog) -> do
-                    putStrLn $ inputFile <> " evaluated; result written to " <> outFilePath
-                    writeFile outFilePath (render $ pretty prog)
-                  Right EvaluationFailure      -> do
-                    -- warn the user that the file failed to evaluate
-                    putStrLn $ inputFile <> " failed to evaluate. Failure written to " <> outFilePath
-                    writeFile outFilePath shownEvaluationFailure
-                  Left _ -> do
-                    -- warn the user that exception is thrown
-                    putStrLn $ "Exception thrown during evaluation of " <> inputFile <>".Written to " <> outFilePath
-                    writeFile outFilePath shownEvaluationFailure
-          )
+      Eval ->
+        for_ inputFiles printAndWrite
+          where printAndWrite inputFile = do
+                  inputTxt <- T.readFile inputFile
+                  let parsed = parseTxt inputTxt
+                      outFilePath = inputFile <> ".expected"
+                  case parsed of
+                    Left (ParseErrorB _) -> do -- specifying parsed to ParserError for the compiler
+                      -- warn the user that the file failed to parse
+                      putStrLn $ inputFile <> " failed to parse. Error written to " <> outFilePath
+                      T.writeFile outFilePath shownParseError
+                    Right pro -> do
+                      res <- try (evaluate $ evalUplcProg (() <$ pro)):: IO (Either SomeException (EvaluationResult UplcProg))
+                      case res of
+                        Right (EvaluationSuccess prog) -> do
+                          putStrLn $ inputFile <> " evaluated; result written to " <> outFilePath
+                          T.writeFile outFilePath (render $ pretty prog)
+                        Right EvaluationFailure      -> do
+                          -- warn the user that the file failed to evaluate
+                          putStrLn $ inputFile <> " failed to evaluate. Failure written to " <> outFilePath
+                          T.writeFile outFilePath shownEvaluationFailure
+                        Left _ -> do
+                          -- warn the user that exception is thrown
+                          putStrLn $ "Exception thrown during evaluation of " <> inputFile <>".Written to " <> outFilePath
+                          T.writeFile outFilePath shownEvaluationFailure
```

Comment:
It's `SomeException` in there, so a `Ctrl-c` gets caught and then swallowed, which is not good.

Perhaps it's time to employ the trick described [here](https://www.schoolofhaskell.com/user/snoyberg/general-haskell/exceptions/catching-all-exceptions) and implemented [here](https://hackage.haskell.org/package/enclosed-exceptions-1.0.3/docs/Control-Exception-Enclosed.html)? It adds some overhead, but an app not responding to Ctrl-c is much more annoying than an app that is a tiny bit slower than what it could be.

Reply:
OK so I pushed https://github.com/input-output-hk/plutus/pull/4688/commits/1dd50eb381ce4c4a9fd5cd7d7943a1732b617875 but I don't know if it does anything. It seems that `ctrl-c` was working before the commit already. But there may be other things that will be good to do this with. If you think of anything I can test it with lmk. Thanks!

Reply:
> It seems that that package is deprecated. They've since written [safe-exceptions](https://github.com/fpco/unliftio/tree/master/unliftio#async-exception-safety) and then they [deprecated it with unliftio](https://www.fpcomplete.com/haskell/library/safe-exceptions/). I'll try that.

`safe-exceptions` and `unliftio` are both a mess in terms of exception handling, unfortunately. The inherent design limitation is that they pretend that only certain types of exceptions are thrown asynchronously, while Haskell allows one to throw any kind of exception asynchronously, so there's this mismatch. Another problem is that both the libraries mask exceptions as uninterruptible in functions like `bracket` and `finally`, which can easily lead to deadlocks (I read a story about that just recently somewhere in our slack, in addition to all the similar stories that I'd heard before that). And on top of that there's this inconsistency where `catch` doesn't mask exceptions as uninterruptible (they're interruptible there like in `base`).

Using a worker thread doesn't have any of these drawbacks.

I do think using `unliftio` is fine for this particular use case. But in general we should be very cautious about exception handling and `unliftio` is not a solution for things that are on the critical path.

Reply:
OK that's good to know. Maybe some of this should go in the style guide? Could you add a section on exception handling?

Reply:
Great idea, created PLT-418.